### ISSUE TITLE
Uno components should be available in web component architecture

### ISSUE THREAD

**Issue by Robbert** [nl-design-system/backlog#100]
## Introduction

The current codebase of the component library that was forked from Uno has a couple of things we would like to change:

- **Use the `nl-` prefix for components and class names**. We would prefer the names to start with `nl-`, so organisations can have a "fork first" policy where they fork the component and change the prefix so they can iterate quickly on existing components without NL Design System becoming the bottleneck. Contributing the modifications back to NL Design System should be a separate step where there is there isn't as much time pressure to accept contributions and quality comes first.
- **CSS Variables for design tokens** instead of hardcoded values for Rijkshuisstijl.
- **Web components** to initiate complex components via the native instead of custom progressive enhancement scripts.
- **ES Modules** instead of System.js
- **Web components** so for many components there is a lesser need to maintain implementations specific to each framework.
- Provide a CSS file that defines CSS variables for the Rijkshuisstijl.

For backwards compatibility with existing Uno class names, it should be relatively to write a SCSS file that makes the current Uno class names an alias for the prefixed BEM class names.

## Components

- Backlink: ([Uno](https://uno.dfront.rijkscloud.nl/componenten/backlink/)) → https://github.com/nl-design-system/backlog/issues/8
- Badge: ([Uno](https://uno.dfront.rijkscloud.nl/componenten/badge/)) - split into three separate components
  - Counter badge: https://github.com/nl-design-system/backlog/issues/7
  - Text badge: https://github.com/nl-design-system/backlog/issues/94
  - Status badge: https://github.com/nl-design-system/backlog/issues/6
  - The color modifiers are specific to Rijkshuisstijl, so for `badge--robijnrood` we will not have a semantic equivalent. More likely we will have `nl-badge--dataviz-1` - `nl-badge--dataviz-20`.
- Button: ([Uno](https://uno.dfront.rijkscloud.nl/componenten/button/)) → #38
  - States: hover / focus / disabled
  - Variants: regular / primary action / submit / next / previous / button with icon
  - Skip for now:
    - busy button: need to investigate accessibilty requirements and usage guidelines, and UX to disable the button
    - link button: we probably want to include that type of styling in a separate link component
    - tertiary: need a more defined use case for this before we can include it
- Card ([Uno](https://uno.dfront.rijkscloud.nl/componenten/card/))
  - mostly skip for now: it is a very complex component, with infinite variants, we can offer little useful information for the generic variant
  - link card: https://github.com/nl-design-system/backlog/issues/101 making an entire section clickable to activate an embedded link is such a widely used pattern we need to include it in an early stage. It might not have to come with any card specific styling.
  - no migration at this stage of these patterns: card grid, card columns
  - no migration at this stage of this variant (or perhaps, a separate component): `list--card-small`
- Chart ([Uno](https://uno.dfront.rijkscloud.nl/componenten/donutchart/))
  - do not migrate, the requirements for many users would be too complex at this stage. The Donut Chart for Duo is quite simple though, we could migrate the current version to `<duo-donut-chart>`
  - we could investigate global design tokens for dataviz colors (needs issue) so they can be used to configure 3rd party libraries in the correct style
  - we could offer a code example how to apply CSS variables for dataviz in third party libraries
- Checkbox: ([Uno](https://uno.dfront.rijkscloud.nl/componenten/checkbox/)) → #64
  - Label for checkbox group vs. label for checkbox item
  - Patterns:
    - Multiple choice pattern: high priority
    - Agree to terms pattern: low priority
    - The choice between radio, checkbox, select: low priority
    - Do not migrate yet: checkbox table, tree checkbox, horizontal checkbox
- Collapse ([Uno](https://uno.dfront.rijkscloud.nl/componenten/collapse/))
  - Expanded icon: https://github.com/nl-design-system/backlog/issues/103
  - Collapsed icon: https://github.com/nl-design-system/backlog/issues/104
  - Accordion: https://github.com/nl-design-system/backlog/issues/102
  - Expandable region https://github.com/nl-design-system/backlog/issues/10
  - Not at this stage: expandable table row
- Combobox ([Uno](https://uno.dfront.rijkscloud.nl/componenten/combobox/))
  - deferred, too complex at the initial stage
- Content ([Uno](https://uno.dfront.rijkscloud.nl/componenten/content/))
  - Container element that styles HTML content from a CMS, so it has side effects and does only style HTML with `nl-` prefixed class names
  - This also depends on basic typography ("broodtekst") components, such as:
    - Article
    - Section
    - Heading 1 - Heading 6
    - Paragraph
    - Intro Paragraph
    - Tables?
    - Lists
  - Do not migrate yet: "Content met inpage" / "Content uitvullen" / "Content extra witruimte" / "Content Header"
- Definition list ([Uno](https://uno.dfront.rijkscloud.nl/componenten/definition-list/))
  - too complex for now: too many use cases with different guidelines
  - the most frequent use case, and the one in most dire need of documentation, is the "data summary" - the table-like overview of form data.
- Footer: ([Uno](https://uno.dfront.rijkscloud.nl/componenten/footer/)) → #56
  - container element met design tokens en richtijnlijnen voor semantische HTML, dat de is de scope van de component
  - "link list" en dergelijke is een losse component, valt nu buiten de scope
  - design token zijn wel belangrijk (kleuren, witruimte)
- File: do not migrate at this stage ([Uno](https://uno.dfront.rijkscloud.nl/componenten/file/))
  - It is a component that is used in many different interfaces, so many organisations would benefit from it. This would be a good candidate for other organisations to contribute at the stage they encounter the need for it.
  - For sake of atomic design, this would be an "Attachment" component, and an "Attachment list"
- Filter ([Uno](https://uno.dfront.rijkscloud.nl/componenten/filter/))
  - this looks like a pattern, do not migrate at this stage
- Header ([Uno](https://uno.dfront.rijkscloud.nl/componenten/header/)) → #55
  - seems like a simple component, and essential in most websites
  - container component, logo should be a separate component, but we could have a `nl-header__logo` slot for this purpose.
  - we should include info about the skip link
  - should investigate what design tokens apply, other than whitespace and color
- Hero image ([Uno](https://uno.dfront.rijkscloud.nl/componenten/hero-image/))
  - Seems like this is the perfect component to start a collaboration with folks with a background in CMS implementations.
  - Not at this stage, wait for use cases.
  - We should include some relevant considerations in the content guidelines though, particulary in regards to contrast of (white) text on top of images.
- Hint [(Uno)](https://uno.dfront.rijkscloud.nl/componenten/hint/)
  - feels like a component that is meant for a very wide range of use cases. From instructions in a form to an polite and informative in-line secondary and tertiary "notification".
  - As white-label component it would be difficult for organisations to style this, because the design tokens would affect all uses cases
  - We don't want to migrate this right now, but implement this as separate components: "form instruction" component, "callout" component
- Image ([Uno](https://uno.dfront.rijkscloud.nl/componenten/images/))
  - This is separate from the "Figure" and "Caption" element
  - Some a11y info is required
  - Seems very convenient to have `.nl-image { width: 100% }` available
  - Some content guidelines would be needed
- Lists ([Uno](https://uno.dfront.rijkscloud.nl/componenten/lists/))
  - "Link list" should be a separate component, because it would disable some features that are generally essential for other list (bullet icons or dashes) and the links should sometimes maximize the clickable area
  - Lists in articles should be available in NL Design System
  - there should be design tokens for white space and icons
  - "Filter list" will not be migrated, that looks like it should be a "Category badge list" component
  - "Task list" or "process steps list" should be a separate component on the backlog, which seems what the "Decimal with background" variant is all about.
- Modal ([Uno](https://uno.dfront.rijkscloud.nl/componenten/modal/))
  - highly complex component, do not migrate at this stage
  - difficult with framework interaction (routing, focus, scrolling into view, small screens)
  - we should include it on the backlog, so we can measure interest
  - we should start collecting User testing feedback about the pattern of using modal dialogs
  - perhaps we can include documentation to discourage the use of this component
- Meta ([Uno](https://uno.dfront.rijkscloud.nl/componenten/meta/))
  - We should have design tokens for this purpose, to have less prominent text in the page while maintaining proper contrast
  - A "meta" component seems too generic, we could include more specific components such as "article metadata paragraph" that use the relevant design tokens
  - But generally these meta components should live in forks for their own specific purposes, and graduate to the central repo when there is more need for this.
- Navigation ([Uno](https://uno.dfront.rijkscloud.nl/componenten/navigation/#))
  - We should probably split this up into several separate components:
    - Navigation bar
    - In-page navigation (table of contents)
    - breadcrumbs
    - prev/next chapter navigation
    - Mega menu
    - Backlink https://github.com/nl-design-system/backlog/issues/8
    - Pagination
- Notification ([Uno](https://github.com/nl-design-system/backlog/issues/9))
- Pagination ([Uno](https://uno.dfront.rijkscloud.nl/componenten/pagination/))
  - also a navigation component
  - should not be difficult to migrate, and many websites have this
  - what design tokens should apply? primary link color?
  - needs next and prev icons
  - needs guidance on maximizing clickable area
  - as with other navigation components, at some point there should be two implementations: one for server-side navigation and one for single page apps. Needs investigation what the API should be for this component to make that work for SPA's.
  - https://github.com/nl-design-system/backlog/issues/106
- Pie chart ([Uno](https://uno.dfront.rijkscloud.nl/componenten/pie-chart/))
  - nope, use a chart library, instructions with what design tokens for dataviz colors
- Quote ([Uno](https://uno.dfront.rijkscloud.nl/componenten/quote/)) → #73
  - separate components, wait for this component to be requested by community
  - at least two separate components: Callout text component and Blockquote component, both with optional attribution
- Radiobutton ([Uno](https://uno.dfront.rijkscloud.nl/componenten/radio/)) → #65
  - see considerations for checkbox component
- Section ([Uno](https://uno.dfront.rijkscloud.nl/componenten/sections/))
  - edit functionality of sections seems very specific, will not migrate
  - wait for use case, very difficult to choose design tokens
  - should it be section-1, section-2, section-3, for different styling levels for different nesting levels, each with their own design tokens
- Skip link ([Uno](https://uno.dfront.rijkscloud.nl/componenten/skiplink/)) → #74
  - seems essential to have
- Spinner ([Uno](https://uno.dfront.rijkscloud.nl/componenten/spinner/))
  - More than a spinner image, the essential information is that we need to provide to developers is how to inform the busy state for parts of the page, and announce when the information is available.
  - Also specifically explain what needs to be done for the spinner in the context of form submissions.
  - Summary: spinner component should probably never be implemented in a page as-is, needs more work than that.
  - Loading pattern should be a separate item in our backlog
  - Also investigate the need for a "Shimmer component" and a "Skeleton component"
  - So maybe, including the spinner component without the relevant extra documentation it might do more harm than good.
- Switch ([Uno](https://uno.dfront.rijkscloud.nl/componenten/switch/)) → #33
  - Do not migrate immediately
  - Deterimine in what contexts this should instead of checkbox / radio
  - Should it always apply in real-time, or is a form submission step acceptable for this?
- Table ([Uno](https://uno.dfront.rijkscloud.nl/componenten/table/))
  - Responsive tables and `<div>` tables would be out of scope for migration within our time constraints, but help is much appreciated
  - Only offer basic styling for tables in the context of an article, and offer design tokens for those
- Tabs ([Uno](https://uno.dfront.rijkscloud.nl/componenten/tabs/))
  - they are discouraged by GOV.UK, we should await 
  - https://uno.dfront.rijkscloud.nl/componenten/tabs/
- Tooltip ([Uno](https://uno.dfront.rijkscloud.nl/componenten/tooltip/))
  - we should probably discourage the use of tooltips
- Timeline ([Uno](https://uno.dfront.rijkscloud.nl/componenten/timeline/)
  - we should probably ask The Hague to contribute this
- Video ([Uno](https://uno.dfront.rijkscloud.nl/componenten/videos/))
  - this components seems to be a figure than a video, we should probably investigate creating a Figure component that works for both
  - content guidelines for video, illustrations, subtitles and transcriptions are available for Rijkshuisstijl, we should see if we can copy those
- Favicon → #61
  - When we have a separate Rijkshuisstijl version of NL Design System, we should include the favicon component there.
[...to be continued]
